res/controllers/Traktor-Kontrol-S3-hid-scripts.js

```diff
@@ -0,0 +1,2227 @@
+            this.trackLoadedHandler();
+        } else {
+            // No track loaded, abort
+            return;
+        }
+    }
+    this.curPosition = value * this.trackDurationSec;
+    this.positionUpdated = true;
+};
+
+TraktorS3.Channel.prototype.vuMeterHandler = function(value) {
+    this.vuMeterUpdated = true;
+    this.vuMeterValue = value;
+};
+
+TraktorS3.Channel.prototype.linkOutputs = function() {
+    this.vuConnection = engine.makeConnection(this.group, "VuMeter", TraktorS3.bind(TraktorS3.Channel.prototype.vuMeterHandler, this));
+    this.clipConnection = engine.makeConnection(this.group, "PeakIndicator", TraktorS3.bind(TraktorS3.Controller.prototype.peakOutput, this.controller));
+    this.controller.linkChannelOutput(this.group, "pfl", TraktorS3.bind(TraktorS3.Controller.prototype.pflOutput, this.controller));
+    for (var j = 1; j <= 8; j++) {
+        this.hotcueCallbacks.push(engine.makeConnection(this.group, "hotcue_" + j + "_enabled",
+            TraktorS3.bind(TraktorS3.Channel.prototype.hotcuesOutput, this)));
+        this.hotcueCallbacks.push(engine.makeConnection(this.group, "hotcue_" + j + "_activate",
+            TraktorS3.bind(TraktorS3.Channel.prototype.hotcuesOutput, this)));
+    }
+};
+
+TraktorS3.Channel.prototype.channelBaseColor = function() {
+    if (this.group === "[Channel4]" && this.controller.channel4InputMode) {
+        return this.controller.hid.LEDColors[this.controller.hid.LEDColors.OFF];
+    }
+    return this.controller.hid.LEDColors[TraktorS3.ChannelColors[this.group]];
+};
+
+// colorOutput drives lights that have the palettized multicolor lights.
+TraktorS3.Channel.prototype.colorOutput = function(value, key) {
+    var ledValue = this.channelBaseColor();
+    if (value === 1 || value === true) {
+        ledValue += TraktorS3.LEDBrightValue;
+    } else {
+        ledValue += TraktorS3.LEDDimValue;
+    }
+    this.controller.hid.setOutput(this.group, key, ledValue, !this.controller.batchingOutputs);
+};
+
+TraktorS3.Channel.prototype.hotcuesOutput = function(_value, group, key) {
+    var deck = this.controller.Channels[group].parentDeck;
+    if (deck.activeChannel !== group) {
+        // Not active, ignore
+        return;
+    }
+    var matches = key.match(/hotcue_(\d+)_/);
+    if (matches.length !== 2) {
+        HIDDebug("Didn't get expected hotcue number from string: " + key);
+        return;
+    }
+    var cueNum = matches[1];
+    deck.lightHotcue(cueNum);
+};
+
+// Returns true if there was an update.
+TraktorS3.Channel.prototype.lightWheelPosition = function() {
+    if (!this.positionUpdated) {
+        return false;
+    }
+    this.positionUpdated = false;
+    var rotations = this.curPosition * (1 / 1.8);  // 1/1.8 is rotations per second
+    // Calculate angle from 0-1.0
+    var angle = rotations - Math.floor(rotations);
+    // The wheel has 8 segments
+    var wheelAngle = 8.0 * angle;
+    var baseLedValue = this.channelBaseColor();
```

Comment:
I'm going to have to try making my S4 Mk3 spin rainbows around the wheels. :laughing:

Reply:
Then make a video with it while isolating a rainbow LED hoop :rainbow:

Reply:
If you want to add an easter egg option for this, that would be cool. Of course you don't have to.

Reply:
that would be very fun, but unfortunately my code assumes 1 color for all the wheel segments... I'll see if there's an easy way to change that, but maybe that's something to save for a subsequent commit :)

Reply:
You could have one spot unlit to indicate the track position.

Reply:
ooo, inverting colors does sound like a good idea.  I'll try it